### PR TITLE
Implement ERC1155Proxy and StaticCallProxy in Solidity

### PR DESCRIPTION
**Pull request by abandeali1**
## Description

- Implements the `ERC1155Proxy` and `StaticCallProxy` in Solidity
- Note: `StaticCallProxy` cannot inherit `IAssetProxy` because its `transferFrom` function is marked as `view`.
- Many revert conditions no longer contain reasons, as they are handled within Solidity
- There are no longer any checks for if the length of `assetData` is a multiple of 32. However, any lengths less than required should fail (TODO: add tests for this)
- Found some pretty bizarre behavior is Solidity's `abi.decode`. When multiple offsets point to the same location in calldata (i.e if `values` == `ids`) then the decoded output is also copied to the same location in memory. This means we cannot modify `values` directly, or we may end up modifying `ids` or `data` as well.
- Thoughts on using `slice` vs `sliceDestructive` here? I opted for `sliceDestructive` because `assetData` is not reused, it is cheaper, and the implementation is arguably simpler.

TODO: gas benchmarking

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: contracts/asset-proxy/contracts/src/ERC1155Proxy.sol [hunk 1]
```diff
@@ -18,349 +18,71 @@
...
-            revert(0, 0)
+        // Decode params from `assetData`
+        // solhint-disable indent
+        (
+            address erc1155TokenAddress,
+            uint256[] memory ids,
+            uint256[] memory values,
+            bytes memory data
+        ) = abi.decode(
+            assetData.sliceDestructive(4, assetData.length),
```
Comment (dorothy-zbornak) [inline contracts/asset-proxy/contracts/src/ERC1155Proxy.sol]: I'm surprised this works with `assetData` being `calldata` and not `memory`. Is it cloning `assetData` to memory at the time of this call?

Comment (abandeali1) [reply contracts/asset-proxy/contracts/src/ERC1155Proxy.sol]: I believe calldata gets copied to memory when you pass it into any internal function (`sliceDestructive` in this case). That might not be implemented for things that require ABIv2 encoding, though (or maybe just structs?).

File: contracts/asset-proxy/contracts/src/StaticCallProxy.sol [hunk 2]
```diff
@@ -18,181 +18,57 @@
...
+            address staticCallTarget,
+            bytes memory staticCallData,
+            bytes32 expectedReturnDataHash
+        ) = abi.decode(
+            assetData.sliceDestructive(4, assetData.length),
+            (address, bytes, bytes32)
+        );
+
+        // Execute staticcall
+        (bool success, bytes memory returnData) = staticCallTarget.staticcall(staticCallData);
```
Comment (dorothy-zbornak) [inline contracts/asset-proxy/contracts/src/StaticCallProxy.sol]: Should we be doing an `extcodesize()` check to safeguard people from accidentally calling EOAs? Could be a problem if people expect a `keccak(bytes(0))` result.

At the very least, we should add a test that confirms that calling an EOA results in `keccak(bytes(0))`.

Comment (abandeali1) [reply contracts/asset-proxy/contracts/src/StaticCallProxy.sol]: I will add a test for this, but it feels a bit excessive to me to add an `extcodesize` or `extcodehash` check here. You really have to know the interface of and trust the `staticCallTarget` beforehand in order to use this.

**Pull request merged**